From b375b4f1131d54315bb9952fcf2eff363b3b29b1 Mon Sep 17 00:00:00 2001
From: Dan Smith <danms@us.ibm.com>
Date: Fri, 29 Jun 2012 09:35:02 -0700
Subject: [PATCH 10/16] Redefine the domain's XML on volume attach/detach

This fixes bug 1004791 by adding new disk definitions to the defined
XML instead of just modifying the running instance.

Cherry picked for stable/essex to fix bug 1012717.

Change-Id: I6596dae7c54158c32bc7b399c55a1797b2d98242
---
 nova/tests/fakelibvirt.py       |  2 ++
 nova/virt/libvirt/connection.py | 14 ++++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/nova/tests/fakelibvirt.py b/nova/tests/fakelibvirt.py
index 24ff5ae..cb63cbe 100644
--- a/nova/tests/fakelibvirt.py
+++ b/nova/tests/fakelibvirt.py
@@ -73,6 +73,8 @@ VIR_DOMAIN_SHUTDOWN = 4
 VIR_DOMAIN_SHUTOFF = 5
 VIR_DOMAIN_CRASHED = 6
 
+VIR_DOMAIN_XML_SECURE = 1
+
 VIR_CPU_COMPARE_ERROR = -1
 VIR_CPU_COMPARE_INCOMPATIBLE = 0
 VIR_CPU_COMPARE_IDENTICAL = 1
diff --git a/nova/virt/libvirt/connection.py b/nova/virt/libvirt/connection.py
index 64e5a2e..da072be 100644
--- a/nova/virt/libvirt/connection.py
+++ b/nova/virt/libvirt/connection.py
@@ -549,6 +549,13 @@ class LibvirtConnection(driver.ComputeDriver):
                                                connection_info,
                                                mount_device)
 
+        # TODO(danms) once libvirt has support for LXC hotplug,
+        # replace this re-define with use of the
+        # VIR_DOMAIN_AFFECT_LIVE & VIR_DOMAIN_AFFECT_CONFIG flags with
+        # attachDevice()
+        domxml = virt_dom.XMLDesc(libvirt.VIR_DOMAIN_XML_SECURE)
+        self._conn.defineXML(domxml)
+
     @staticmethod
     def _get_disk_xml(xml, device):
         """Returns the xml for the disk mounted at device"""
@@ -583,6 +590,13 @@ class LibvirtConnection(driver.ComputeDriver):
                                       connection_info,
                                       mount_device)
 
+        # TODO(danms) once libvirt has support for LXC hotplug,
+        # replace this re-define with use of the
+        # VIR_DOMAIN_AFFECT_LIVE & VIR_DOMAIN_AFFECT_CONFIG flags with
+        # detachDevice()
+        domxml = virt_dom.XMLDesc(libvirt.VIR_DOMAIN_XML_SECURE)
+        self._conn.defineXML(domxml)
+
     @exception.wrap_exception()
     def _attach_lxc_volume(self, xml, virt_dom, instance_name):
         LOG.info(_('attaching LXC block device'))
-- 
1.7.11.4

